Add file printer for JIT backend use.

vuzelac-cadence opened the pull request:
Introduce host based file printer for backend JIT use.
It allows for concurrent writes to the same files and also avoids running into problems on windows with NTDLL failing on fprintf from JIT code.

Related to #4392, #3845 

Use like this in libjit.cpp for example:
```
#include <glow/LLVMIRCodeGen/JITFilePrinter.h>
extern "C" {
extern JITFileWriterIF _jitFileWriter;
}
```
Actual calls:
`_jitFileWriter.fopen();`

Testing: #4392 needs to provide testing, this MR will include replace libjit.cpp I/O calls, and the tests should pass.

opti-mix commented:
@vuzelac-cadence I wonder if the same functionality could have been sold using simpler methods, e.g. something described in #2991. One could have exposed in the host process custom versions of certain functions or expose some new APIs with the required functionality. WDYT?

opti-mix commented on include/glow/LLVMIRCodeGen/JITFilePrinter.h:
```diff
@@ -0,0 +1,45 @@
+  std::unordered_map<FILE *, std::string> fileMap_;
+  std::condition_variable readyCV;
+  std::mutex mtx_;
+};
```
New line is missing.

opti-mix commented on include/glow/LLVMIRCodeGen/JITFilePrinter.h:
```diff
@@ -0,0 +1,45 @@
+#include <cstdarg>
+#include <unordered_map>
+
+struct JITFileWriterIF {
```
Could you add Doxygen comments for the types and their members?

opti-mix commented on include/glow/LLVMIRCodeGen/LLVMIRGen.h:
```diff
@@ -371,6 +371,8 @@ class LLVMIRGen {
   /// Generates LLVM IR that materializes the string literal \p str.
   virtual llvm::Value *emitStringConst(llvm::IRBuilder<> &builder,
                                        llvm::StringRef str);
+  /// Generate pointer to the host side file printer for JIT code to use. \p F.
```
The comment looks a bit strange at the end. I mean `\p F`. BTW, how is F used in the function?

opti-mix commented on lib/LLVMIRCodeGen/JITFilePrinter.cpp:
```diff
@@ -0,0 +1,73 @@
+  return ret;
+}
+
+void LLVMIRGen::generateJITFileWriter(llvm::Function *F) {
```
Please add comments explaining what this function does and how it is implemented.

opti-mix commented on lib/LLVMIRCodeGen/JITFilePrinter.cpp:
```diff
@@ -0,0 +1,73 @@
+  llvm::Constant *writerPtr = llvm::ConstantExpr::getIntToPtr(
+      writerV, llvm::PointerType::getUnqual(ptrType));
+  writer->setInitializer(writerPtr);
+}
```
New line missing.

opti-mix commented on lib/LLVMIRCodeGen/JITFilePrinter.cpp:
```diff
@@ -0,0 +1,73 @@
+
+int JITFileWriter::JITfclose(FILE *f) {
+  std::lock_guard<std::mutex> lck(jitfileWriter.mtx_);
+  std::string file = jitfileWriter.fileMap_[f];
```
Don't you need to free the entry in fileMap_ afterwards?

Also don't you need to check if `f` is even in `fileMap_`?

vuzelac-cadence replied:
Fixed both

vuzelac-cadence commented:
@opti-mix thanks for a quick review. I've just send out the commit we had internally to supplement my comments in  #4392 . Will cleanup.

vuzelac-cadence commented:
@opti-mix,  I was aware of 1st two approaches you mentioned in #2991. Now, using 
llvm::sys::DynamicLibrary::AddSymbol("global_counter", &counter);
certainly looks very sleek and I prefer it much more. I'll see to add my printer using this one.

ayermolo commented:
@vuzelac-cadence thanks for up streaming. :D
Related issue:
https://github.com/pytorch/glow/issues/3937

vuzelac-cadence commented:
@opti-mix , regarding` llvm::sys::AddSymbol`

My idea was not to export the `_jitFileWriter` but a pointer to it, thinking that will allow me to continue using the `weak `attribute on the JIT side in order to be able to check if Host has provided the printer:

const JITFileWriterIF *_jitFileWriter __attribute__((weak)) = nullptr;

But defining the symbol  using `AddSymbol` will not override the above symbol anymore.
Now I can explore more, but was thinking - who cares ?  JIT can still use use whatever printer it wants (through some other selection mechanism). We actually don't need this selection . I don't see the need for selection in #4392 either. Do you anticipate a need for such weak symbol definition ?.

opti-mix commented on lib/LLVMIRCodeGen/LLVMBackend.cpp:
```diff
@@ -99,6 +99,8 @@ void LLVMBackend::emitJitMain(LLVMIRGen &irgen) const {
   // Terminate the function.
   builder.CreateRetVoid();
+  // Emit JIT file printer.
+  irgen.generateJITFileWriter(func);
```
LLVMBackend is used sometimes for AOT compilation, not just for a JIT. Thus, registering the JITFileWriter unconditionally could be not such a good idea in general. But I don't have a strong opinion here.

vuzelac-cadence replied:
I think of this "registering" as simply "providing". Backend is free not to use it. 
Note that we need CPU backend provided dumps as much as Xtensa dumps, throughout our compile and use flow. Therefore I moved registering here. We originally had a backend option to enabled this, however, we settled with the current approach merely due to it being one-liner, which allows for easier rebase until (and if ever) we upstream.

opti-mix commented on lib/LLVMIRCodeGen/JITFilePrinter.cpp:
```diff
@@ -0,0 +1,73 @@
+/// Expose host side file printer for JIT code to use.
+/// The function simply exposes the writer object as a global symbol
+/// that will be avaiable for JIT to use.
+void LLVMIRGen::generateJITFileWriter(llvm::Function *F) {
```
It seems like you don't use the `F` parameter at all. What is the purpose of `F` in a general case?

opti-mix replied:
Could you address this comment?

vuzelac-cadence commented:
@opti-mix I agree `HostEnvironment `is the ideal solution. But, we were trying to solve problem on hands and the approach looked not too patchy, if you think of it as a part of the JIT tensor dump infrastructure and nothing more. 
We don't have bandwidth to provide better implementation. We are ok if this is not upstreamed - the original intent of this MR is to backup the issues I brought up in #4392 . As is, #4392 will be buggy.
Note that we do have many host provided structures and functions we emit into Xtensa JIT, from Glow Xtensa Backend, which matches your expectation that backend should take care of this. However, this one is special as we need it for CPU Backend as well.

opti-mix commented:
@vuzelac-cadence Yeah, I see your point. And I really don't have a very strong opinion here. 

@jfix71 What do you think? Is this feature in its current form worth being upstreamed or do you think it will always stay very backend-specific or use-case specific?

jfix71 commented:
Also no strong feelings. Agreed it's not the most general/ideal solution. But it seems potentially useful and pretty well self-contained, so I'd be fine with upstreaming it.

opti-mix commented:
@vuzelac-cadence OK, let's land it. Could you please address the remaining comments and formatting issues?

vuzelac-cadence commented:
@opti-mix , @jfix71 , format check fails on these files:

	modified:   torch_glow/src/PyTorchModelLoader.cpp
	modified:   torch_glow/tests/nodes/rsub_test.py

I don't have them in my commit. I see they are indeed not formatted.

jfix71 commented:
@vuzelac-cadence Yes, they are broken in master. Will get it fixed soon.

vuzelac-cadence commented:
@jfix71 , rebased to re-trigger the build. This is ready now

jfix71 commented on lib/LLVMIRCodeGen/JITFilePrinter.cpp:
```diff
@@ -0,0 +1,73 @@
+  if (!jitfileWriter.fileMap_.count(f)) {
+    return -1;
+  }
+  std::string file = jitfileWriter.fileMap_[f];
```
nit: `const std::string &file`
